File: metrics_sql/tax_1_py/joining.py
import pandas as pd
import sqlite3
import time
import os
import psutil
from memory_profiler import memory_usage

def sql_joining_task():
    total_lines = 30
    core_lines = 6  

    process = psutil.Process(os.getpid())
    cpu_start = process.cpu_times().user
    wall_start = time.time()

    # Connect to SQLite DB
    conn = sqlite3.connect("benchmark.db")

    # Load main CSV
    df_main = pd.read_csv("../birth_gp_ratios.csv")
    df_main.to_sql("birth_gp_ratios", conn, if_exists="replace", index=False)

    # Create dummy lookup
    dummy_lookup = pd.DataFrame({
        "gss_code": df_main["gss_code"].dropna().unique()[:10],
        "region": [f"Region_{i+1}" for i in range(10)]
    })
    dummy_lookup.to_sql("dummy_lookup", conn, if_exists="replace", index=False)

    # SQL Join
    query = """
    SELECT b.*, l.region
    FROM birth_gp_ratios b
    LEFT JOIN dummy_lookup l ON b.gss_code = l.gss_code
    """
    joined_df = pd.read_sql_query(query, conn)
    joined_df.to_csv("joined_data_sql.csv", index=False)

    conn.close()

    wall_end = time.time()
    cpu_end = process.cpu_times().user
    runtime = round(wall_end - wall_start, 3)
    cpu_runtime = round(cpu_end - cpu_start, 3)

    print(f"Runtime (wall): {runtime} seconds")
    print(f"CPU Runtime: {cpu_runtime} seconds")
    print(f"Total Lines of Code: {total_lines}")
    print(f"Core Lines of Code: {core_lines}")

    return runtime, cpu_runtime, total_lines, core_lines

if __name__ == "__main__":
    mem_usage = memory_usage((sql_joining_task,), max_iterations=1)
    mem_used = round(max(mem_usage) - min(mem_usage), 3)
    print(f"Memory Usage: {mem_used} MB")

File: metrics_r/taxonomy_1/joining.R
# --------------------- SETUP (Not Benchmarked) ---------------------
setwd("C:/Users/HP/Downloads/metrics_r/taxonomy_1")

# Load required package
if (!requireNamespace("pryr", quietly = TRUE)) {
  install.packages("pryr")
}
library(pryr)

# Load data
df_main <- read.csv("../birth_gp_ratios.csv")

# Track total lines of code in script
script_path <- "joining.R"
if (file.exists(script_path)) {
  total_lines_of_code <- length(readLines(script_path))
} else {
  total_lines_of_code <- NA
  warning("⚠️ Script file not found. LOC count failed.")
}

# --------------------- BENCHMARK STARTS HERE ---------------------
start_time <- Sys.time()
cpu_start <- proc.time()
mem_before <- mem_used()

# --- Core Task: Create dummy lookup and perform join ---
gss_codes <- unique(na.omit(df_main$gss_code))[1:10]
dummy_lookup <- data.frame(
  gss_code = gss_codes,
  region = paste0("Region_", seq_along(gss_codes))
)

merged <- merge(df_main, dummy_lookup, by = "gss_code", all.x = TRUE)
write.csv(merged, "joined_data.csv", row.names = FALSE)

# --------------------- BENCHMARK ENDS HERE ---------------------
mem_after <- mem_used()
cpu_end <- proc.time()
end_time <- Sys.time()

# Compute benchmark metrics
runtime <- round(as.numeric(difftime(end_time, start_time, units = "secs")), 3)
cpu_time <- round((cpu_end - cpu_start)[["user.self"]], 3)
mem_used_mb <- round((mem_after - mem_before) / 1024^2, 3)

# Estimate lines used for core task: 3 lines
core_task_lines <- 3

# Output results
cat("✅ Performed left join with dummy region lookup on gss_code\n")
cat("Runtime (seconds):", runtime, "\n")
cat("CPU Time (seconds):", cpu_time, "\n")
cat("Memory usage (MB):", mem_used_mb, "\n")
cat("Total Lines of Code (script_path):", total_lines_of_code, "\n")
cat("Lines of Core Task Only:", core_task_lines, "\n")
